### PR TITLE
#162414184 Implement track comment history feature

### PR DESCRIPTION
**Pull request by Maxfurry**
#### What does this PR do?
Tracks history of comments and threads

#### Description of Task to be completed?
- After a comment is updated or deleted
- It should be possible to get all updated or deleted comments
- All users updated or deleted comments
- All comment in an article updated or deleted comments
- All threads in a comment updated or deleted comments

#### How should this be manually tested?
- Clone this repository
- Open your terminal and put in git checkout ft-track-comment-history-#162414184
to checkout to this branch
- Run npm install to install all dependencies
- Start your local server 
- Insert a user and article into your database
- Create a comment

#### What are the relevant pivotal tracker stories?
PT Link: https://www.pivotaltracker.com/story/show/162414184

#### Screenshots
screen shot 2019-02-07 at 2 55 13 pm

### CODE REVIEW

File: controllers/articlesCommentController.js [hunk 1]
```diff
@@ -53,7 +63,7 @@ async function findCommentController(req, res) {
     });
     return successResponse(res, COMMENT_SUCCESS_RETURN_MESSAGE, OK_CODE, comment);
   } catch (error) {
-    return failureResponse(res, SERVER_ERROR_MESSAGE, SERVER_ERROR_CODE);
+    return failureResponse(res, error, SERVER_ERROR_CODE);
```
Comment (codrex) [inline controllers/articlesCommentController.js]: Why are you sending `error` to the client?

File: controllers/articlesCommentController.js [hunk 2]
```diff
@@ -21,12 +25,18 @@ const Comment = require('../lib/modelManagers/articlesComment'),
...
+    const comments = await Comment.createComment({
+      comment,
+      userId,
+      articleId: id,
+      mainCommentId
+    });
     return successResponse(res, COMMENT_CREATED, RESOURCE_CREATED_CODE, comments);
   } catch (error) {
-    return failureResponse(res, SERVER_ERROR_MESSAGE, SERVER_ERROR_CODE);
+    return failureResponse(res, error, SERVER_ERROR_CODE);
```
Comment (codrex) [inline controllers/articlesCommentController.js]: error ??

File: controllers/articlesCommentController.js [hunk 3]
```diff
@@ -85,8 +95,101 @@ async function findCommentThreadController(req, res) {
...
+ * @param {*} req
+ * @param {*} res
+ * @returns {*} *
+ */
+async function updateCommentController(req, res) {
+  try {
+    const userId = req.decodedToken.payLoad;
+    const { commentId, comment } = req.body;
+    const { articleId } = req.params;
+    const commenter = await Comment.findComment(articleId, commentId);
```
Comment (codrex) [inline controllers/articlesCommentController.js]: `commenter`: The value return from `findComment` should be a comment and not a commenter. Please always use descriptive variable names.

File: lib/modelManagers/commentsHistoryModel.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+class CommentsHistoryManager {
+  /**
+   * @static
+   * @memberof ArticlesComments
+   * @returns {object} The query result.
+   */
+  static async getAllHistory() {
+    const commentHistory = await ArticlesComments.findAndCountAll({
+      where: {
+        $or: [{ updated: true }, { deleted: true }]
```
Comment (codrex) [inline lib/modelManagers/commentsHistoryModel.js]: What happens to comments that were neither updated or deleted?

File: lib/modelManagers/commentsHistoryModel.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+   * @memberof ArticlesComments
+   * @param {string} columnName The field to search
+   * @param {string} value THe value to search for in the field
+   * @returns {object} The query result.
+   */
+  static async getHistory(columnName, value) {
+    const commentHistory = await ArticlesComments.findOne({
+      where: {
+        [columnName]: value,
+        $or: [{ updated: true }, { deleted: true }]
```
Comment (codrex) [inline lib/modelManagers/commentsHistoryModel.js]: same as ☝️

File: routes/api/articles.js [hunk 6]
```diff
@@ -297,4 +297,13 @@ router.get(
   commentController.findCommentThreadController
 );
 
+// Update a comment
```
Comment (codrex) [inline routes/api/articles.js]: remove inline comments

File: routes/api/articles.js [hunk 7]
```diff
@@ -297,4 +297,13 @@ router.get(
...
+router.put('/:articleId/comments/', jwtValidator, commentController.updateCommentController);
+
+// Get all comment history for an article
```
Comment (codrex) [inline routes/api/articles.js]: Make comment history its own controller

File: controllers/articlesCommentController.js (hunk 3 above)
Comment (codrex) [reply controllers/articlesCommentController.js]: You don't have to find comment here since the same comment will be returned from the operation below.
```js
await Comment.updateArticleComment({ updated: true }, commentId);

```

File: controllers/articlesCommentController.js [hunk 8]
```diff
@@ -85,8 +95,101 @@ async function findCommentThreadController(req, res) {
...
+    await Comment.updateArticleComment({ updated: true }, commentId);
+    const createComment = await Comment.createComment({
+      comment,
+      userId,
+      articleId,
+      mainCommentId: commenter.mainCommentId,
```
Comment (codrex) [inline controllers/articlesCommentController.js]: Please explain why you have this ` mainCommentId: commenter.mainCommentId,`?

File: controllers/articlesCommentController.js [hunk 9]
```diff
@@ -85,8 +95,101 @@ async function findCommentThreadController(req, res) {
...
+  }
+}
+
+/**
+ * @description Gets comment history under a specific article Controller
+ * @param {*} req
+ * @param {*} res
+ * @returns {*} *
+ */
+async function getArticlesHistory(req, res) {
```
Comment (codrex) [inline controllers/articlesCommentController.js]: Please explain why we have `getArticlesHistory`?

File: controllers/articlesCommentController.js (hunk 3 above)
Comment (Maxfurry) [reply controllers/articlesCommentController.js]: I need to find comment because I need to check if it has a mainCommentId so that I can update the threads under it with the new commentId has their parent

File: lib/modelManagers/commentsHistoryModel.js (hunk 4 above)
Comment (Maxfurry) [reply lib/modelManagers/commentsHistoryModel.js]: They don't get returned
I only want to get the comments or threads that were changed

File: routes/api/articles.js (hunk 7 above)
Comment (Maxfurry) [reply routes/api/articles.js]: Noted

File: controllers/articlesCommentController.js (hunk 8 above)
Comment (Maxfurry) [reply controllers/articlesCommentController.js]: The mainCommentId refers the parent thread, if it is not updated with the new parent id, it will not show up under the modified comment because it still tied to the old comment

File: controllers/articlesCommentController.js (hunk 9 above)
Comment (Maxfurry) [reply controllers/articlesCommentController.js]: Get all comments updated or deleted comments from an article

File: controllers/articlesCommentController.js (hunk 1 above)
Comment (Maxfurry) [reply controllers/articlesCommentController.js]: I am not done, still testing the code.
I have already corrected it

File: lib/modelManagers/commentsHistoryModel.js (hunk 4 above)
Comment (codrex) [reply lib/modelManagers/commentsHistoryModel.js]: If that is the case, name the function to reflect what the function is doing.

Comment (Maxfurry) [reply lib/modelManagers/commentsHistoryModel.js]: Noted

File: controllers/articlesCommentController.js [hunk 10]
```diff
@@ -85,8 +93,59 @@ async function findCommentThreadController(req, res) {
...
+ * @param {*} req
+ * @param {*} res
+ * @returns {*} *
+ */
+async function updateCommentController(req, res) {
+  try {
+    const userId = req.decodedToken.payLoad;
+    const { commentId, comment } = req.body;
+    const { articleId } = req.params;
+    const getCommentDetails = await Comment.findComment(articleId, commentId);
```
Comment (codrex) [inline controllers/articlesCommentController.js]: `const getCommentDetails = await Comment.findComment(articleId, commentId);` I don't think you need to perform this operation. `updateArticleComment` will return a comment object. It will be better you make use of that.

File: lib/modelManagers/commentsHistoryModel.js [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * Comment History Manager Class
+ */
+class CommentsHistoryManager {
+  /**
+   * @static
+   * @memberof ArticlesComments
+   * @returns {object} The query result.
+   */
+  static async getAllmodifiedComments() {
```
Comment (codrex) [inline lib/modelManagers/commentsHistoryModel.js]: Use camelcase for variable names. `getAllmodifiedComments`

File: controllers/articlesCommentController.js (hunk 3 above)
Comment (codrex) [reply controllers/articlesCommentController.js]: The above line of code also returns the same record you are getting from the find comment function.

File: controllers/articlesCommentController.js (hunk 8 above)
Comment (codrex) [reply controllers/articlesCommentController.js]: ok

File: controllers/articlesCommentController.js (hunk 2 above)
Comment (Maxfurry) [reply controllers/articlesCommentController.js]: fixed

File: routes/api/articles.js (hunk 6 above)
Comment (Maxfurry) [reply routes/api/articles.js]: corrected

File: lib/modelManagers/commentsHistoryModel.js (hunk 5 above)
Comment (Maxfurry) [reply lib/modelManagers/commentsHistoryModel.js]: Fixed

File: lib/modelManagers/commentsHistoryModel.js (hunk 11 above)
Comment (Maxfurry) [reply lib/modelManagers/commentsHistoryModel.js]: Fixed

File: controllers/articlesCommentController.js [hunk 12]
```diff
@@ -57,7 +70,7 @@ async function findCommentController(req, res) {
       return failureResponse(res, COMMENT_NOT_FOUND, NOT_FOUND_CODE);
     }
     const comment = comments.map((thread) => {
-      thread.threads = thread.threads.length;
+      thread.dataValues.threads = thread.threads.length;
```
Comment (codrex) [inline controllers/articlesCommentController.js:73]: `thread.dataValues.threads` why the change here?

Comment (Maxfurry) [reply controllers/articlesCommentController.js:73]: I added dataValues because without dataValues it was not giving the desired result. It was Lucky that implemented it, but when you said we would get the same result without dataValues, Kabir removed it. When I merged with mine it was not having the same effect that was there before, so I had to reverse it.

Comment (codrex) [reply controllers/articlesCommentController.js:73]: Why is this so? It worked well for @Luckzman and @ibraheemkabir. Why is it different for you?

Comment (Maxfurry) [reply controllers/articlesCommentController.js:73]: dataValues was there when lucky used it, Ibrahim didn't use it he only removed all dataValues in the file. He had no use of it.

Comment (Maxfurry) [reply controllers/articlesCommentController.js:73]: Without dataValues
screen shot 2019-02-07 at 7 42 50 pm

Comment (Maxfurry) [reply controllers/articlesCommentController.js:73]: With dataValues (Desired Result)
screen shot 2019-02-07 at 7 42 31 pm

Comment (Maxfurry) [reply controllers/articlesCommentController.js:73]: The thread count is where the effect is seen

Comment (codrex) [reply controllers/articlesCommentController.js:73]: Cool

**Review by codrex — APPROVED** [review]
